sound/soc/sof/intel/hda-stream.c

```diff
@@ -122,19 +137,28 @@ int hda_dsp_stream_setup_bdl(struct snd_sof_dev *sdev,
 	ioc = sdev->hda->no_ipc_position ?
 	      !stream->no_period_wakeup : 0;
 
-	for (i = 0; i < periods; i++) {
-		if (i == (periods - 1) && remain)
-			/* set the last small entry */
-			offset = hda_setup_bdle(sdev, dmab,
-						stream, &bdl, offset,
-						remain, 0);
+	for (i = 0; i < entries; i++) {
+		bytes = period_bytes - offset % period_bytes;
+		if (is_wrap)
+			bytes = min(bytes, init_offset - offset);
 		else
-			offset = hda_setup_bdle(sdev, dmab,
-						stream, &bdl, offset,
-						period_bytes, ioc);
+			bytes = min(bytes, stream->bufsize - offset);
+
+		/* only set ioc at period bundary */
+		set_ioc = (offset + bytes) % period_bytes ? 0 : 1;
+		offset = hda_setup_bdle(sdev, dmab,
+					stream, &bdl, offset,
+					bytes, set_ioc && ioc);
+
+		if (offset == stream->bufsize) {
+			offset = 0;
+			is_wrap = true;
+		}
+
+		size += bytes;
 	}
```

Comment:
I am unable to determine if the code is correct or not, it's significantly changed from the initial code and I am hesitant to sign-off here. Maybe it's correct, but it's not self-explanatory nor easy to debug/maintain. It'd be easier to deal with logical steps, such as from init_offset to first period start, then all regular periods, then end of buffer alignment, then wrap until last period and again remaining bytes?

Reply:
@plbossart I've verified that the logic here is correct. quite impressive @keyonjie 

But might be easier to follow if you break it up like pierre suggests.

Reply:
hmm, actually I spent hours to abstract and make it as simple as what it is now, otherwise, it will be inevitable lengthy and low efficiency with those 5 steps Pierre mentioned.

Reply:
You may have spent hours but I have no way of knowing if it works or not. 
How many times have we faced issues with corner cases in this sort of code?
Please.

Reply:
So, shall I try add more comments with logic analyzing?

Reply:
The ask is to find a way to make things self-explanatory with all corner cases handled. You can do this with simple code or comments, your choice.